### ISSUE TITLE
Support FDE via LUKS & Multiple LVM Volumes

### ISSUE THREAD

**Issue by That-Random-Guy** [calamares/calamares#1128]
**Is your feature request related to a problem? Please describe.**
AFAIK, Calamares _does_ support LUKS/LVM but does not currently allow a Full-Disk Encryption (FDE) setup where one wishes to have multiple LVM volumes under a single LUKS container.

**Describe the solution you'd like**
I'd like for Calamares to allow for a FDE setup where I can have more than one partition encrypted via LUKS. To be more specific, I would like it if I was able to create an LUKS container, and then be able to create the proper partitions under the encrypted LVM PV. Currently, Calamares only supports creation of LVM volumes if a PV is already manually provisioned. This process should be supported officially via the installer so that users aren't expected to provision their own partitions (via terminal) and can simply use the installer to set this up on the fly. Having this supported, FDE that isn't solely implemented with a standard non-LVM partition is allowed. When selecting "Erase disk" and checking off encryption, LVM is not involved; therefore there currently remains no visible/clear option to achieve the desired setup.

**Describe alternatives you've considered**
I have attempted to both provision LUKS and LVM together before running Calamares where I have received install failures for all attempts.

- [**First attempt**: ] I created a LUKS1 container using cryptsetup, unlocked it, and provisioned basic LVM on the container. I then ran Calamres, which correctly identified the partitions but failed to then create the specified, distinct /, home, swap volumes.
- [**Second attempt:**  ] I provisioned a basic LVM config, ran Calamares, and then individually created the /, home, swap volumes while checking off the encryption check box (thus converting them all to LUKS partitions). 

Both cases failed for obvious reasons (at least to me).

**Additional context**
I would like to point out that this is something I've only tested using Lubuntu's 19.04 image.

I think the ideal would be something RH/Fedora's Anaconda supports or implements. The reason I wanted to install Lubuntu is because I find it to fit my needs for my older laptop. Finding that it's not possible to get my desired setup, I decided to test Fedora's LXQt (which is one of the main reasons I chose Lubuntu) and it installed itself as I was hoping to install Lubuntu. I am not happy stuck using Fedora but it will suffice for the moment. My hope is that a feature evolves in the installer to allow for LVM under LUKS as it does for the 18.04 image Lubuntu offers (which I tested).

When I say volumes, I am referring to logical volumes under LVM.

**Comment by abucodonosor** [comment]
Almost all *Ubuntu(s) and/or Debian based Distributions do not support full LUKS 
encryption in a normal setup , so I doubt what you want will (ever) work there.

Also LVM is kind not really done yet , in general.

However can you provide full logs from these failed installations?

Also ping @wxl and @tsimonq2

**Comment by abucodonosor** [comment]
Also @cjlcarvalho and @stikonas may know *if* LVM implementation can do that at the moment.

**Comment by tsimonq2** [comment]
> Almost all *Ubuntu(s) and/or Debian based Distributions do not support full LUKS
encryption in a normal setup with Calamares

I have no idea what you're talking about. We have this working in 19.04. The way @That-Random-Guy is describing it is how Ubuntu does it with Ubiquity and probably how Fedora does it with Anaconda, except with /boot unencrypted and outside of the LVM volume.

**Comment by kkofler** [comment]
Yes, Anaconda defaults to LVM, so encryption defaults to LVM over LUKS. And indeed, Anaconda does not support encrypted `/boot`.

**Comment by abucodonosor** [comment]
> ... except with /boot unencrypted and outside of the LVM volume.

Which is not *full* encryption , is it ?

**Comment by kkofler** [comment]
It's the quasi-full encryption Anaconda always uses. Calamares may well be the only installer out there that supports unlocking `/boot` from GRUB.

**Comment by stikonas** [comment]
I suspect that in manual partitioning you can get Luks+lvm setup working. But untested. And definitely not in automatic mode.

On 25 April 2019 11:25:32 BST, Kevin Kofler <notifications@github.com> wrote:
>It's the quasi-full encryption Anaconda always uses. Calamares may well
>be the only installer out there that supports unlocking `/boot` from
>GRUB.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/calamares/calamares/issues/1128#issuecomment-486616635

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by abucodonosor** [comment]
@kkofler 

I don't talk about Anaconda , the way LUKS is implemented there is different.

The way LUKS is implemented in Calamares kind need all to be encryted.
With unencrypted /boot you also can just disable LUKS, having your key on
unencrypted /boot does not make any sense.
( You explained it already in #1073 I think )

**Comment by kkofler** [comment]
Yes, I explained it already, and yes, what you wrote is correct. Calamares does not properly support unencrypted `/boot`, that is known and already filed elsewhere. But the default is to encrypt `/boot` anyway, so that is also what would be implemented for the LVM over LUKS case in Calamares if we were to support it, unlike the implementation in Anaconda.

**Comment by That-Random-Guy** [comment]
I realize my desired setup may not technically be considered FDE, but I thought the perquisite was to have the bootloader unencrypted for all implementations. I wasn't aware of different means. I referenced the ManualFullSystemEncryption wiki page on Ubuntu where it described having an ESP and boot partition where boot is encrypted inside of the container. Since my system uses UEFI, I felt like this would be possible under Ubiquity but I didn't think it was also necessary for Calamares (since I don't know what it's doing).

I am including the logs & info that I receive for the failure of the first described attempt where a single LUKS container is utilized and later provisioned with LVM. I thought it was _supposed_ to fail because Calamares isn't aware of my passphrase/key. The second attempt, in my mind, should also have failed---and it did---because I didn't provision the distinct partitions correctly. Those are simply guesses, however.
[calamares_stdout.txt](https://github.com/calamares/calamares/files/3121497/calamares_stdout.txt)
[session.log](https://github.com/calamares/calamares/files/3121498/session.log)

In case this wasn't clear before, this is **only** a feature request. I'm **not** demanding anything nor do I expect this to be supported. If the team doesn't want to pick this up just yet, I'd prefer they close the issue.

I've been adamant towards the described setup but if there's a different way to have FDE with multiple partitions, then I'd be _equally_ as interested.

I do not currently have a substantial reason for why LVM under LUKS should be supported but I suppose it boils down to having/wanting the flexibility LVM offers.

If this type of feature is not inline with the project, then I'm good with that.

What I'd be interested in having confirmed is if Calamares does support some form of FDE?

From what's been mentioned so far, can I not simply encrypt **/** via the installer and use that single partition for the system? Why would boot need to be encrypted and would the same apply to ESP?

**Comment by abucodonosor** [comment]
@That-Random-Guy 

I guess lots other bugs need be fixed until we get to LUKS :)

While trying to look at this bug I found some other issues ( still have some to report )

You can do that sure but Calamares implementation will put you key into the initrd
which is stored in /boot .. Having unencrypted /boot to store your key for the 
encrypted rootfs is nonsense since that setup is not what you expect to be from
security point of view.

**Comment by That-Random-Guy** [comment]
> You can do that sure but Calamares implementation will put you key into the initrd
> which is stored in /boot .. Having unencrypted /boot to store your key for the
> encrypted rootfs is nonsense since that setup is not what you expect to be from
> security point of view.

Is there currently a different way to do that then or should I just wait? Not sure what to do.

**Comment by abucodonosor** [comment]
@That-Random-Guy 

No from Calamares.. But you can install with Calamaes then change the setup to
unlock with systemd or some GUI asking you on boot the partition(s) password.
( You need ask maintainers of the Distributions you want to install the 
best way of doing that..)
